src/data-sources/components/DataSourceForm.tsx

```diff
@@ -70,15 +71,45 @@ const DataSourceForm = ( { children, onSave }: DataSourceFormProps ) => {
 	const steps = Children.toArray( children );
 	const singleStep = steps.length === 1 || screen === 'editDataSource';
 
-	const stepHeadings = [ 'Setup', 'Scope' ];
+	const getStepHeadings = () => {
+		const headings = [ __( 'Setup' ) ]; // First step is always Setup
+
+		if (
+			Children.toArray( children ).some(
+				child =>
+					isValidElement( child ) &&
+					( child.type === DataSourceForm.Scope || child.type === 'Scope' )
+			)
+		) {
+			headings.push( __( 'Scope' ) );
+		}
+
+		if (
+			Children.toArray( children ).some(
+				child =>
+					isValidElement( child ) &&
+					( child.type === DataSourceForm.Blocks || child.type === 'Blocks' )
+			)
+		) {
+			headings.push( __( 'Blocks' ) );
+		}
+
+		return headings;
+	};
+
+	const stepHeadings = getStepHeadings();
 
 	const canProceedToNextStep = () => {
```

Comment:
Curious, why is this a function (with no args) instead of just a variable computed in the render scope?

Reply:
[This reply]( https://github.com/Automattic/remote-data-blocks/pull/322#discussion_r1936237909) can work here too